artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPSessionCallback.java

```diff
@@ -418,6 +418,17 @@ public void cancel(Object brokerConsumer, Message message, boolean updateCounts)
       }
    }
 
+   /** to be used when credits were drained between handle and actual delivery */
+   public void returnToQueue(Object brokerConsumer, MessageReference reference) {
```

Comment:
Is this issue re proton credits possibly occuring in earlier versions say 2.21? Asking as had odd issue with consumers just pausing like they run out of credits.

Reply:
The issue here is really about a memory leak. 

I had credits when handleDelivery was called. later on I called deliver, the AMQPStandardMessage would be referenced on a DeliveryImpl that would stay forever in the Link.workingList until the session or the connection is closed.


I had other issues that I fixed as part of this PR regarding QueueConsumer. where the iterator would hung in the Iterators. Perhaps that one would cause some issues.

https://github.com/apache/activemq-artemis/pull/4375/commits/a1aea9a79aa844673ef705d2a60fadb4435accfc#diff-fc4a9a0dc01f79ecc1435246dc85446fa57d6d04e1dffed53c522d8c145623fdR3098-R3105

Reply:
This is being tested on ConnectionLeakTest::testCheckIterators*, which is using https://github.com/check-leak/check-leak

Reply:
@michaelandrepearce those issues we had were mostly related to closing consumers with a lot of pre-fetch in the queue, which is an anti-pattern. One user reported issues here, and even though we will make sure he changes his usage,I  still wanted to fix these.

Although the Iterator and Journal were issues independent of prefetch or AMQP.